framework/src/org/checkerframework/common/jaifinference/JaifInferenceUtils.java

```diff
@@ -0,0 +1,485 @@
+     */
+    private static List<AnnotationMirror> annotationsToIgnore =
+            new ArrayList<AnnotationMirror>();
+
+    /**
+     * @param classSymbol is the symbol representing the class containing the method.
+     * @param methodElt is the method element.
+     * @param atf is the annotated type factory.
+     * @return the ATM of methodElt in the .jaif file of the class with symbol
+     * classSymbol, or null if it can't be found.
+     */
+    public static AnnotatedTypeMirror getMethodReturnTypeInJaif(
+            ClassSymbol classSymbol, ExecutableElement methodElt,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return null; // Static block.
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, new AScene());
+            if (clazz == null) return null;
+            String methodName = JVMNames.getJVMMethodName(methodElt);
+            AMethod method = clazz.methods.get(methodName);
+            if (method != null && method.returnType != null) {
+                Set<Annotation> prevAnnos = method.returnType.tlAnnotationsHere;
+                return setOfAnnotationsToATM(prevAnnos, atf, false,
+                            methodElt.getReturnType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * @param classSymbol is the symbol representing the class containing the method.
+     * @param fieldNode is the node representing the field.
+     * @param atf is the annotated type factory.
+     * @return the ATM of fieldNode in the .jaif file of the class with symbol
+     * classSymbol, or null if it can't be found.
+     */
+    public static AnnotatedTypeMirror getFieldTypeInJaif(
+            ClassSymbol classSymbol, FieldAccessNode fieldNode,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return null; // Static block.
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+
+        AScene scene = new AScene();
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, scene);
+            if (clazz == null) return null;
+            String fieldName = TreeUtils.getFieldName(fieldNode.getTree());
+            AField field = clazz.fields.get(fieldName);
+            if (field == null) {
+                return null;
+            }
+            Set<Annotation> annos = field.tlAnnotationsHere;
+            if (annos != null && annos.size() > 0) {
+                return setOfAnnotationsToATM(annos, atf, false, fieldNode.getType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * Updates the type of the field lhs in the .jaif file of the class with
```

Comment:
This alternate approach is more efficient.

The con is that we would have to rely on the SourceChecker.shutdownHook() method to do something at the end of the whole type-checking process (to confirm, look for the email sent to the CF dev list with subject "Calling typeProcessingOver() or similar when there are warnings").

I tried to implement this alternate approach: SourceChecker.shutdownHook() is called on a new thread. For some reason, this new thread does not have the AFU.jar in its classpath, so I keep getting ClassNotDef exceptions. I don't know where the classpath must be changed to fix this.

Alternatively, I could write scenes into .jaif files after type-checking each compilation unit. If type-checking a compilation unit modified more than one .jaif file, I would have to write in more than one .jaif files. This approach would be less efficient than the one you suggested, but better than the current one, and does not rely on the shutdownHook() method.

Please, let me know what you think is best.